cats/views.py

```diff
@@ -0,0 +1,154 @@
+
+from django.contrib.auth.decorators import login_required
+from django.core.mail import send_mail
+from django.core.paginator import Paginator
+from django.shortcuts import render, get_object_or_404, redirect
+from django.template.loader import render_to_string
+from django.urls import reverse
+from django.views.generic import TemplateView, ListView, DetailView
+
+from cats.forms import RentalForm, SearchForm
+from cats.models import Cat, Species, Breed, Rental
+
+
+class IndexView(TemplateView):
+    """Simple index view"""
+    template_name = 'cats/index.html'
+
+
+class AboutView(TemplateView):
+    """Simple about view"""
+    template_name = 'cats/about.html'
+
+
+class SpeciesListView(ListView):
+    """View to choose a Species"""
+    model = Species
+    template_name = 'cats/species.html'
+
+
+def explore_list(request):
+    """Lists Cats from all kinds of Species"""
+    cats = None
+    page_obj = None
+    search_form = SearchForm(request.GET or None)
+
+    """
+    If form is valid, show list.
+    If form is not valid, SearchForm will show proper hint
+    """
+    if search_form.is_valid():
+        date_from = search_form.cleaned_data['date_from']
+        date_to = search_form.cleaned_data['date_to']
+        cats = Cat.objects.filter_by_dates(date_from, date_to)
+
+        paginator = Paginator(cats, 5)
+        page_number = request.GET.get('page')
+        page_obj = paginator.get_page(page_number)
+
+    context = {'cats_list': cats, 'search_form': search_form, 'page_obj': page_obj}
+    return render(request, 'cats/explore_list.html', context)
+
+
+def cats_list(request, species_id):
+    """Lists Cats from only one Species"""
+    cats = None
+    page_obj = None
+    search_form = SearchForm(request.GET or None)
+    breeds = Breed.objects.all()
+
+    # If such species id do not exists, then make species=None (html will show proper hint then)
+    try:
+        species = Species.objects.get(id=species_id)
+    except Species.DoesNotExist:
+        species = None
+
+    # If form is valid, show list. If not, SearchForm will show proper hint
+    if search_form.is_valid():
+        date_from = search_form.cleaned_data['date_from']
+        date_to = search_form.cleaned_data['date_to']
+        species_cats = Cat.objects.filter(breed__species=species_id)
+        cats = species_cats.filter_by_dates(date_from, date_to)
+
+        paginator = Paginator(cats, 5)
+        page_number = request.GET.get('page')
+        page_obj = paginator.get_page(page_number)
+
+    context = {
+        'cats_list': cats,
+        'breeds_list': breeds,
+        'species_id': species_id,
+        'search_form': search_form,
+        'species': species,
+        'page_obj': page_obj
+    }
+    return render(request, 'cats/cats_list.html', context)
+
+
+class CatDetailView(DetailView):
+    """Detail view for a Cat"""
+    model = Cat
+    template_name = 'cats/details.html'
+
+
+@login_required
+def rental_dates(request, cat_id):
```

Comment:
Why not `FormVIew`?

Reply:
Honestly, I'm not sure when should I use class-based or function-based views.

Courses I took were saying - always try to use class-based views, as they're more good-looking in code and they does some work for you.
Other ones said - class-based views are mostly not needed to be used and they'll just give you less control and less understaning of your code ^^

So I've just used both options as I wished :P

Reply:
Generally, CBV is preferred way, probably the only still existing use case for FBV is when you need to process multiple forms in one view.

Reply:
Thanks, that's probably the clearest answer I've read about choosing between them 😛